src/tools/crawler/crawl.rs

```diff
@@ -0,0 +1,121 @@
+//! Contains structs and methods to crawl ripple network according to instruction at https://xrpl.org/peer-crawler.html
+
+use std::{
+    net::{IpAddr, SocketAddr},
+    num::ParseIntError,
+    time::Duration,
+};
+
+use reqwest::{Client, StatusCode};
+use serde::Deserialize;
+use thiserror::Error;
+use tokio::time::Instant;
+
+const CRAWLER_DEFAULT_PORT: u16 = 51235;
+const CRAWLER_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+pub(super) struct Peer {
+    pub(super) complete_ledgers: Option<String>,
+    pub(super) complete_shards: Option<String>,
+    pub(super) ip: Option<String>,
+    pub(super) port: Option<Port>,
+    pub(super) public_key: String,
+    #[serde(rename = "type")]
+    pub(super) connection_type: String,
+    pub(super) uptime: u32,
+    pub(super) version: String,
+}
+
+impl Peer {
+    /// Returns port number for the peer. If no port was present in the response then returns the default port.
+    /// Returns [ParseIntError] if the response contained invalid value.
+    pub(super) fn port(&self) -> Result<u16, ParseIntError> {
+        self.port
+            .as_ref()
+            .map(|p| match p {
+                Port::Number(n) => Ok(*n),
+                Port::String(s) => s.parse(),
+            })
+            .unwrap_or(Ok(CRAWLER_DEFAULT_PORT))
+    }
+}
+
+#[derive(Deserialize)]
+pub(super) struct CrawlResponse {
+    pub(super) overlay: Overlay,
+    pub(super) server: Server,
+}
+
+#[derive(Deserialize)]
+pub(super) struct Overlay {
+    pub(super) active: Vec<Peer>,
+}
+
+#[allow(dead_code)]
+#[derive(Deserialize)]
+pub(super) struct Server {
+    pub(super) build_version: String,
+    pub(super) server_state: String,
+    pub(super) uptime: u32,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub(super) enum Port {
+    Number(u16),
+    String(String),
+}
+
+/// Connects to `https://IP:PORT/crawl` to query `addr's` peers.
+/// Returns (response, time it took to connect) on success, or
+/// [CrawlError] when unable to connect.
+pub(super) async fn get_crawl_response(
+    addr: SocketAddr,
+) -> Result<(CrawlResponse, Duration), CrawlError> {
+    let url = format!("https://{}:{}/crawl", format_ip_for_url(addr), addr.port());
+    let client = Client::builder()
```

Comment:
how often is this function expected to be called? it might be preferable for the `Client` to persist if it happens quite often

Reply:
Done. It required some more changes but it's better now, thanks.